l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -0,0 +1,733 @@
+    /// @dev If the L2 deposit finalization transaction fails, the `_refundRecipient` will receive the `_l2Value`.
+    /// Please note, the contract may change the refund recipient's address to eliminate sending funds to addresses
+    /// out of control.
+    /// - If `_refundRecipient` is a contract on L1, the refund will be sent to the aliased `_refundRecipient`.
+    /// - If `_refundRecipient` is set to `address(0)` and the sender has NO deployed bytecode on L1, the refund will
+    /// be sent to the `msg.sender` address.
+    /// - If `_refundRecipient` is set to `address(0)` and the sender has deployed bytecode on L1, the refund will be
+    /// sent to the aliased `msg.sender` address.
+    /// @dev The address aliasing of L1 contracts as refund recipient on L2 is necessary to guarantee that the funds
+    /// are controllable through the Mailbox, since the Mailbox applies address aliasing to the from address for the
+    /// L2 tx if the L1 msg.sender is a contract. Without address aliasing for L1 contracts as refund recipients they
+    /// would not be able to make proper L2 tx requests through the Mailbox to use or withdraw the funds from L2, and
+    /// the funds would be lost.
+    /// @return l2TxHash The L2 transaction hash of deposit finalization.
+    function depositLegacyErc20Bridge(
+        address _prevMsgSender,
+        address _l2Receiver,
+        address _l1Token,
+        uint256 _amount,
+        uint256 _l2TxGasLimit,
+        uint256 _l2TxGasPerPubdataByte,
+        address _refundRecipient
+    ) external payable override onlyLegacyBridge nonReentrant whenNotPaused returns (bytes32 l2TxHash) {
+        require(_l1Token != L1_WETH_TOKEN, "ShB: WETH deposit not supported 2");
+
+        bytes32 _assetId;
+        bytes memory l2BridgeMintCalldata;
+
+        {
+            // Inner call to encode data to decrease local var numbers
+            _assetId = _ensureTokenRegisteredWithNTV(_l1Token);
+
+            // solhint-disable-next-line func-named-parameters
+            l2BridgeMintCalldata = abi.encode(
+                _amount,
+                _prevMsgSender,
+                _l2Receiver,
+                getERC20Getters(_l1Token),
+                _l1Token
+            );
+        }
+
+        {
+            bytes memory l2TxCalldata = l1AssetRouter._getDepositL2Calldata(
+                ERA_CHAIN_ID,
+                _prevMsgSender,
+                _assetId,
+                l2BridgeMintCalldata
+            );
+
+            // If the refund recipient is not specified, the refund will be sent to the sender of the transaction.
+            // Otherwise, the refund will be sent to the specified address.
+            // If the recipient is a contract on L1, the address alias will be applied.
+            address refundRecipient = AddressAliasHelper.actualRefundRecipient(_refundRecipient, _prevMsgSender);
+
+            L2TransactionRequestDirect memory request = L2TransactionRequestDirect({
```

Comment:
Here the L2AssetRouter can receive the call from multiple addresses on L1, which is not ideal, we should send it through the main L1AssetRouter I think maybe?